### PR TITLE
[BEAM-3273] Make InMemoryArtifactStagerService more patient

### PR DESCRIPTION
**Pull request by tgroh**
Wait until there are no outstanding putArtifact requests before
completing the call to commitManifest, and then check to ensure that the
two agree in advance.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand:
   - [x] What the pull request does
   - [x] Why it does it
   - [x] How it does it
   - [ ] Why this approach
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by tgroh** [comment]
retest this please

**Comment by jkff** [comment]
Still has a checkstyle violation and didn't proceed past that.

**Comment by tgroh** [comment]
Should be fixed now, sorry about that.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java [hunk 1]
```diff
@@ -43,26 +48,46 @@
...
 
   @Override
   public void commitManifest(
       ArtifactApi.CommitManifestRequest request,
       StreamObserver<ArtifactApi.CommitManifestResponse> responseObserver) {
     this.manifest = request.getManifest();
-    responseObserver.onNext(CommitManifestResponse.getDefaultInstance());
-    responseObserver.onCompleted();
+    while (activePuts.get() != 0) {
+      // Wait for existing calls to complete.
```
Comment (jkff) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java]: Can there be new calls after this loop?

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java [hunk 2]
```diff
@@ -117,6 +142,8 @@ public void onCompleted() {
               writer.stream.toByteArray());
         } catch (NoSuchAlgorithmException e) {
           throw new AssertionError("The Java Spec requires all JVMs to support MD5", e);
+        } finally {
+          activePuts.getAndDecrement();
```
Comment (jkff) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java]: What if `writer == null`?

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java (hunk 1 above)
Comment (tgroh) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java]: Yes; however, I generally expect that tests (which is where this should be used) will be sufficiently responsive to 

I can add a AtomicBoolean committing and throw if I a put tries to start after we start committing. Should I?

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java (hunk 2 above)
Comment (tgroh) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java]: Done.

**Comment by tgroh** [comment]
PTAL

**Comment by tgroh** [comment]
On second thought, I'm trying to solve the wrong problem entirely.

The current `Map` is a `HashMap` instead of a `ConcurrentHashMap`, and since the server can (does) execute in a background thread, we have visibility issues.

Instead of the original change, I've modified this to only force that the uploaded artifacts is a `ConcurrentMap`.

I will need to add a separate change surrounding the actual relation of `CommitManifest`, but ultimately yes, the client should receive a positive ack from the service that an artifact has been staged before attempting to commit a manifest which contains that artifact

**Comment by tgroh** [comment]
(PTAL, the new change is much simpler and certainly more obviously reasonable)

**Comment by tgroh** [comment]
I ran 10000 instances of the test and got no failures with the new change and at least one without it as verification that this is in fact the issue,.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java [hunk 3]
```diff
@@ -70,7 +77,12 @@ public void commitManifest(
   }
 
   public Manifest getManifest() {
-    return manifest;
+    return manifest.get();
+  }
+
+  public void clear() {
```
Comment (jkff) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java]: This is a new method that isn't called. Did you mean to include some more files in this PR?

Comment (tgroh) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/InMemoryArtifactStagerService.java]: Oh no I have failed to cleanup the temporary change that allowed me to run the test many times.

RM'd

**Pull request merged**